code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java

```diff
@@ -170,6 +176,39 @@ protected void modifyCompute(long hostId, Instance instance, boolean add) {
         objectManager.persist(host);
     }
 
+    protected void modifyDisk(long hostId, Instance instance, boolean add) {
+        CacheManager cm = CacheManager.getCacheManagerInstance(this.objectManager);
+        InstanceInfo instanceInfo = cm.getInstanceInfoForHost(hostId, instance.getId());
+
+        for (Entry<String, InstanceDiskReserveInfo> entry : instanceInfo.getAllReservedDisksInfo()) {
+            String diskDevicePath = entry.getKey();
+            InstanceDiskReserveInfo reserveInfo = entry.getValue();
+            Long reserveSize = reserveInfo.getReservedSize();
+
+            // figure out available size
+            DiskInfo diskInfo = cm.getDiskInfoForHost(hostId, diskDevicePath);
+            Long allocated = diskInfo.getAllocatedSize();
+            if (add && (diskInfo.getCapacity() - allocated >= reserveSize)) {
+                diskInfo.addAllocatedSize(reserveSize);
+                log.info("allocated disk space on disk [{}] with total = {}, {} {} {} = {} as used",
+                        diskInfo.getDiskDevicePath(), diskInfo.getCapacity(), allocated, "+", reserveSize,
+                        allocated + reserveSize);
+
+                // mark reserved info for this instance
+                reserveInfo.setAllocated(true);
+            } else if (!add && reserveInfo.isAllocated()
+                    && allocated >= reserveSize) {
```

Comment:
Same thing here: what happens when !add and allocated < reservedSize?

We should _probably_ raise an error, but I'd also consider just setting allocated back to zero. Thoughts on which is better?